firestore/FirestoreSwiftUIExample/ViewModels/RestaurantListViewModel.swift

```diff
@@ -47,12 +47,15 @@ class RestaurantListViewModel: ObservableObject {
         guard let self = self else { return }
         self.restaurants = documents.compactMap { document in
           do {
-            return try document.data(as: Restaurant.self)
+            var restaurant = try document.data(as: Restaurant.self)
+            restaurant?.reference = document.reference
```

Comment:
There's a bit of trickiness going on here where the document reference is supported by the decoder, but it won't decode the document reference into a parameter of type `DocumentReference?` due to the optional. But we can't make the document reference nonnull on `Restaurant`, because the restaurant values we're working with don't necessarily exist on the server. Adding the optional document reference parameter directly to the `Restaurant` type increases its flexibility, but at the cost of increased complexity--when we do need to use the `reference` parameter, we can't be sure of its existence without examining the value directly, and when we don't, it requires extra cruft in initialization (such as in this example).

Fortunately, with granular types and generics, we can use a separate type to represent a context that we can drop the restaurant into. In this case, the context is 'a value that has a corresponding Firestore document', which we can represent like so:

```swift
struct FirestoreValue<T: Codable, Identifiable>: Identifiable {
  let value: T
  let snapshot: QueryDocumentSnapshot

  init(snapshot: QueryDocumentSnapshot) throws {
    self.snapshot = snapshot
    self.value = try snapshot.data(as: T.self)
  }
}
```

This way you can wrap a `Restaurant` in a `FirestoreValue` whenever it's instantiated as a result of a Firestore query, which lets you separate the `Restaurant`s that were created on the client from the ones that came from the server and allows you to access the Firestore bits directly when you need them. Since Firestore-backed `Restaurant`s and non-Firestore-backed `Restaurant`s are now representable as distinct types (`FirestoreValue<Restaurant>` vs `Restaurant`), we can use them in method parameters to explicitly specify whether or not we need to access the document reference, and the compiler will throw type errors if we're unable to.

Making this change is optional, since it'll require going back and refactoring some infrastructural code, so don't feel like you have to refactor this unless you're in the mood to wrangle the type system.

Reply:
Great info. Thanks for the explanation. I'll give it a shot!